File: app/src/main/java/com/aspark/carebuddy/view/user/UserHomeActivity.java
package com.aspark.carebuddy.view.user;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.aspark.carebuddy.Contract;
import com.aspark.carebuddy.R;
import com.aspark.carebuddy.databinding.ActivityUserHomeBinding;
import com.aspark.carebuddy.model.LocationServicesProvider;
import com.aspark.carebuddy.model.UserModel;
import com.aspark.carebuddy.presenter.UserHomePresenter;
import com.aspark.carebuddy.view.MainActivity;
import com.aspark.carebuddy.view.MapActivity;

public class UserHomeActivity extends AppCompatActivity implements View.OnClickListener {

    boolean isSignedIn = true;
    ActivityUserHomeBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityUserHomeBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setUserSignedIn();

        //TODO currentUser is null when app is opened after destroying it.
        Log.i("UserHomeActivity", "onCreate: currentUser: "+ UserModel.getCurrentUser());

        ActivityResultLauncher<String[]> locationPermissionRequest =
                registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), result -> {

                    Boolean fineLocationGranted = result.getOrDefault(Manifest.permission.ACCESS_FINE_LOCATION,false);
                    Boolean coarseLocationGranted = result.getOrDefault(Manifest.permission.ACCESS_COARSE_LOCATION,false);

                  //  LocationServicesProvider locationServicesProvider = new LocationServicesProvider(this);

                    if(fineLocationGranted != null && fineLocationGranted) {

                        Toast.makeText(this, "Fine location granted", Toast.LENGTH_SHORT).show();
                        //locationServicesProvider.getLastLocation(this, mapView, savedInstanceState);

                        Intent intent = new Intent(this,MapActivity.class);
                        startActivity(intent);
                        finish();

                    }
                    else if( coarseLocationGranted !=null && coarseLocationGranted) {

                        Toast.makeText(this, "Coarse Location Granted", Toast.LENGTH_SHORT).show();
                    }
                    else {
                        Toast.makeText(this, "Location Permission denied", Toast.LENGTH_SHORT).show();


                    }
                });

//        locationPermissionRequest.launch(new String[]{
//                Manifest.permission.ACCESS_FINE_LOCATION,
//                Manifest.permission.ACCESS_COARSE_LOCATION
//        });

        if(ContextCompat.checkSelfPermission(this,Manifest.permission.ACCESS_FINE_LOCATION) ==
                PackageManager.PERMISSION_GRANTED) {

            Toast.makeText(this, "Checked that permission was GRANTED", Toast.LENGTH_SHORT).show();
        }
        else {

            Toast.makeText(this, "onCreate: checked that permission was DENIED", Toast.LENGTH_SHORT).show();
            locationPermissionRequest.launch(new String[]{
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.ACCESS_COARSE_LOCATION
            });
        }

        binding.signOutBtn.setOnClickListener( view -> {

            Log.w("UserHomeActivity", "onCreate: User Logged Out" );

            Intent intent = new Intent(this, MainActivity.class);
            startActivity(intent);

            isSignedIn = false;
            setUserSignedIn();
            finish();
        });

        binding.locationIcon.setOnClickListener(this);
        binding.homeText.setOnClickListener(this);
        binding.downArrowIcon.setOnClickListener(this);
        binding.bookServiceBtn.setOnClickListener(view -> {

            Log.d("UserHomeActivity", "bookService: button clicked ");
            Contract.Presenter.PresenterUserHome presenterUserHome = new UserHomePresenter();
            presenterUserHome.bookService(view.getContext());

        });


    }

    private void setUserSignedIn() {

        SharedPreferences preferences = getSharedPreferences(getPackageName(),MODE_PRIVATE);
        SharedPreferences.Editor editor = preferences.edit();

        Log.i("UserHomeActivity", "setUserSignedIn: isSignedIn "+isSignedIn);

        editor.putBoolean("isSignedIn",isSignedIn);

        if (isSignedIn)
               editor.putString("username",UserModel.getCurrentUser().getEmail());
        else
               editor.putString("username",null);

        editor.apply();
    }


    @Override
    public void onClick(View view) {




    }
}

File: app/src/main/java/com/aspark/carebuddy/view/user/UserHomeActivity.kt
package com.aspark.carebuddy.view.user

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.aspark.carebuddy.UserHomeViewModel
import com.aspark.carebuddy.databinding.ActivityUserHomeBinding
import com.aspark.carebuddy.map.MapActivity
import com.aspark.carebuddy.model.User
import com.aspark.carebuddy.view.MainActivity

class UserHomeActivity : AppCompatActivity() {

    private var isUserSignedIn = true
    private lateinit var binding : ActivityUserHomeBinding
    private val viewModel: UserHomeViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityUserHomeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setIsUserSignedIn()

        Log.i("UserHomeActivity", "onCreate: currentUser: " + User.currentUser)

        //check if notification permission granted
        if (ContextCompat.checkSelfPermission(this,
            Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED)
            Log.i("UserHomeActivity", "onCreate: Checked notification permission granted")

        else if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)){

            //show a educational ui to inform the user of benefits of accepting this permission
            Toast.makeText(this, "Accept notification permission to not " +
                    "miss any update about your bookings",
                Toast.LENGTH_LONG).show()
        }

        else {
            Log.e("UserHomeActivity", "onCreate: Checked notification permission was denied " )
            showNotificationPermissionDialog()
        }

        //check if location permission granted
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
            Log.i("UserHomeActivity", "onCreate: Checked location permission granted")

        else {
            Log.e("UserHomeActivity", "onCreate: Checked location permission was denied " )
            showLocationPermissionDialog()
        }

        binding.signOutBtn.setOnClickListener {

            Log.w("UserHomeActivity", "onCreate: User Logged Out")
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)

            isUserSignedIn = false
            setIsUserSignedIn()
            finish()
        }

//        binding.locationIcon.setOnClickListener(this)
//        binding.homeText.setOnClickListener(this)
//        binding.downArrowIcon.setOnClickListener(this)

        binding.bookServiceBtn.setOnClickListener {

            Log.d("UserHomeActivity", "bookService: button clicked ")
            viewModel.bookServiceClickListener()
        }
    }

    private fun showNotificationPermissionDialog(){

        val notificationPermissionRequest = registerForActivityResult(
            ActivityResultContracts.RequestPermission()){ isGranted: Boolean ->

            if (isGranted){

                Log.d("UserHomeActivity", "showNotificationPermissionDialog: " +
                        "notification permission granted ")

            }
            else{

                Log.e("UserHomeActivity", "showNotificationPermissionDialog: " +
                        "notification permission denied")
            }
        }

        notificationPermissionRequest.launch(Manifest.permission.POST_NOTIFICATIONS)
    }

    private fun showLocationPermissionDialog() {

        val locationPermissionRequest = registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()) { result: Map<String, Boolean?> ->

            val fineLocationGranted =
                result.getOrDefault(Manifest.permission.ACCESS_FINE_LOCATION, false)

            val coarseLocationGranted =
                result.getOrDefault(Manifest.permission.ACCESS_COARSE_LOCATION, false)

            if (fineLocationGranted != null && fineLocationGranted) {

                Log.d("UserHomeActivity", "onCreate: fine Location Granted ")
                val intent = Intent(this, MapActivity::class.java)
                startActivity(intent)
                finish()

            } else if (coarseLocationGranted != null && coarseLocationGranted)
                Log.d("UserHomeActivity", "onCreate: Coarse Location Granted ")

            else Log.d("UserHomeActivity", "onCreate: Location Permission Denied ")

        }

        locationPermissionRequest.launch(
            arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION)
        )
    }

    private fun setIsUserSignedIn() {

        val preferences = getSharedPreferences(packageName, MODE_PRIVATE)
        val editor = preferences.edit()

        Log.i("UserHomeActivity", "setIsUserSignedIn: isUserSignedIn $isUserSignedIn")
        editor.putBoolean("isUserSignedIn", isUserSignedIn)

        if (isUserSignedIn) editor.putString("userEmail", User.currentUser.email)
        else editor.putString("userEmail", null)

        editor.apply()
    }
}